iron-input.html

```diff
@@ -12,300 +12,417 @@
+         * will be rejected. The recommended format should be a list of allowed characters,
+         * for example, `[a-zA-Z0-9.+-!;:]`.
+         *
+         * This pattern represents the allowed characters for the field; as the user inputs text,
+         * each individual character will be checked against the pattern (rather than checking
+         * the entire value as a whole). If a character is not a match, it will be rejected.
+         *
+         * Pasted input will have each character checked individually; if any character
+         * doesn't match `allowedPattern`, the entire pasted string will be rejected.
+         *
+         * Note: if you were using `iron-input` in 1.0, you were also required to
+         * set `prevent-invalid-input`. This is no longer needed as of Polymer 2.0,
+         * and will be set automatically for you if an `allowedPattern` is provided.
+         *
+         */
+        allowedPattern: {
+          type: String
+        },
+
+        /**
+         * Set to true to auto-validate the input value as you type.
+         */
+        autoValidate: {
+          type: Boolean,
+          value: false
+        }
+      },
 
-    behaviors: [
-      Polymer.IronValidatableBehavior
-    ],
+      observers: [
+        '_bindValueChanged(bindValue, _inputElement)'
+      ],
 
-    properties: {
+      listeners: {
+        'input': '_onInput',
+        'keypress': '_onKeypress'
+      },
+
+      registered: function() {
+        // Feature detect whether we need to patch dispatchEvent (i.e. on FF and IE).
+        if (!this._canDispatchEventOnDisabled()) {
+          this._origDispatchEvent = this.dispatchEvent;
+          this.dispatchEvent = this._dispatchEventFirefoxIE;
+        }
+      },
+
+      created: function() {
+        Polymer.IronA11yAnnouncer.requestAvailability();
+        this._previousValidInput = '';
+        this._patternAlreadyChecked = false;
+      },
+
+      attached: function() {
+        // If the input is added at a later time, update the internal reference.
+        this._observer = Polymer.dom(this).observeNodes(function(info) {
+          this._initSlottedInput();
+        }.bind(this));
+      },
+
+      detached: function() {
+        Polymer.dom(this).unobserveNodes(this._observer);
+        this._observer = null;
+      },
 
       /**
-       * Use this property instead of `value` for two-way data binding.
+       * Returns the distributed <input> element.
        */
-      bindValue: {
-        observer: '_bindValueChanged',
-        type: String
+      get inputElement () {
+        return this._inputElement;
+      },
+
+      _initSlottedInput: function() {
+        this._inputElement = this.getEffectiveChildren()[0];
+
+        if (this.inputElement && this.inputElement.value) {
+          this.bindValue = this.inputElement.value;
+        }
+
+        this.fire('iron-input-ready');
```

Comment:
This is not fired by ready. This is needed because the `input` is distributed, and parents (ahem `paper-input`) need to know when it's _actually_ ready. Should I rename it?